generators/chipyard/src/main/scala/ChipTop.scala

```diff
@@ -0,0 +1,102 @@
+ */
+abstract class BaseChipTop()(implicit val p: Parameters) extends RawModule with HasTestHarnessFunctions {
+
+  // A publicly accessible list of IO cells (useful for a floorplanning tool, for example)
+  val iocells = ArrayBuffer.empty[IOCell]
+  // A list of functions to call in the test harness
+  val harnessFunctions = ArrayBuffer.empty[TestHarnessFunction]
+  // The system clock
+  val systemClock = Wire(Input(Clock()))
+  // The system reset (synchronous to clock)
+  val systemReset = Wire(Input(Bool()))
+
+  // The system module specified by BuildSystem
+  val system = withClockAndReset(systemClock, systemReset) { p(BuildSystem)(p) }
+
+  // Call all of the IOBinders and provide them with a default clock and reset
+  withClockAndReset(systemClock, systemReset) {
+    val (_ports, _iocells, _harnessFunctions) = p(IOBinders).values.map(_(system)).flatten.unzip3
+    // We ignore _ports for now...
+    iocells ++= _iocells.flatten
+    harnessFunctions ++= _harnessFunctions.flatten
+  }
+
+}
+
+/**
+ * A simple clock and reset implementation that punches out clock and reset ports with the same
+ * names as the implicit clock and reset for standard Module classes. Reset is synchronous to
+ * clock, which may not be a good idea to use for tapeouts.
+ */
+trait HasChipTopSimpleClockAndReset { this: BaseChipTop =>
+
+  val (clock, systemClockIO) = IOCell.generateIOFromSignal(systemClock, Some("iocell_clock"))
+  val (reset, systemResetIO) = IOCell.generateIOFromSignal(systemReset, Some("iocell_reset"))
+
+  iocells ++= systemClockIO
+  iocells ++= systemResetIO
+
+  // Add a TestHarnessFunction that connects clock and reset
+  harnessFunctions += { (th: TestHarness) => {
+    // Connect clock; it's not done implicitly with RawModule
+    clock := th.clock
+    // Connect reset; it's not done implicitly with RawModule
+    // Note that we need to use dutReset, not harnessReset
+    reset := th.dutReset
+    Nil
+  } }
+
+}
+
+/**
+ * Variant of HasChipTopSimpleClockAndReset that adds a reset synchronizer so that the top-level reset
+ * can be asynchronous with clock, which is useful for tapeout configs.
+ */
+trait HasChipTopSimpleClockAndCaughtReset { this: BaseChipTop =>
+
+  val asyncResetCore = Wire(Input(Bool()))
+  systemReset := ResetCatchAndSync(systemClock, asyncResetCore)
+
+  val (clock, systemClockIO) = IOCell.generateIOFromSignal(systemClock, Some("iocell_clock"))
+  val (areset, asyncResetIO) = IOCell.generateIOFromSignal(asyncResetCore, Some("iocell_areset"))
+
+  iocells ++= systemClockIO
+  iocells ++= asyncResetIO
+
+  // Add a TestHarnessFunction that connects clock and areset
+  harnessFunctions += { (th: TestHarness) => {
+    // Connect clock; it's not done implicitly with RawModule
+    clock := th.clock
+    // Connect reset; it's not done implicitly with RawModule
+    // Note that we need to use dutReset, not harnessReset
+    areset := th.dutReset
+    Nil
+  } }
+
+}
```

Comment:
Is there a way to use the Config system to specify whether to use ChipTop or ChipTopCaughtReset? This smells similar to the movement towards parameterized traits across all our projects. The duplicated code across these two traits seems especially suspicious.

Reply:
We could make a `ChipTopReset` Field and just change behavior inside of `HasChipTopSimpleClockAndReset`, but I did want an example of creating your own `ChipTop`. Probably not a big deal either way, but I agree having to change `TOP` is super annoying. Once @colinschmidt or whoever is able to migrate barstools to stage/phase, we can make it easy to have `TOP` be an annotation input rather than a command line flag (related to @abejgonzalez's comment above)

Reply:
That's a good point. We can just leave these as two separate traits then, for educational purposes.

Reply:
OK, sounds good. I actually did make the change and pushed to the `iocell_notwotops` branch if you want to look at it.